storage/sql/coresql/wrapper/wrapper.go

```diff
@@ -0,0 +1,110 @@
+	"database/sql"
+
+	"github.com/google/trillian/storage"
+)
+
+// TreeWrapper provides SQL wrapping for raw tree storage.
+type TreeWrapper interface {
+	GetTreeRevisionIncludingSize(tx *sql.Tx, treeID, treeSize int64) (int64, int64, error)
+	GetSubtrees(tx *sql.Tx, treeID, treeRevision int64, nodeIDs []storage.NodeID, subtreeScanFn func(*sql.Rows) error) error
+	// SetSubtrees args should be a 4 tuple of (treeID, prefix, subtreeBytes, writeRevision) for each new subtree
+	SetSubtrees(tx *sql.Tx, args []interface{}) error
+}
+
+// LogStatementProvider provides SQL statement objects for log storage.
+type LogStatementProvider interface {
+	GetActiveLogsStmt(tx *sql.Tx) (*sql.Stmt, error)
+	GetActiveLogsWithWorkStmt(tx *sql.Tx) (*sql.Stmt, error)
+	DeleteUnsequencedStmt(tx *sql.Tx, num int) (*sql.Stmt, error)
+	GetLeavesByIndexStmt(tx *sql.Tx, num int) (*sql.Stmt, error)
+	GetLeavesByMerkleHashStmt(tx *sql.Tx, num int, orderBySequence bool) (*sql.Stmt, error)
+	GetLeavesByLeafIdentityHashStmt(tx *sql.Tx, num int) (*sql.Stmt, error)
+	InsertTreeHeadStmt(tx *sql.Tx) (*sql.Stmt, error)
+	GetLatestSignedLogRootStmt(tx *sql.Tx) (*sql.Stmt, error)
+	GetQueuedLeavesStmt(tx *sql.Tx) (*sql.Stmt, error)
+	InsertUnsequencedEntryStmt(tx *sql.Tx) (*sql.Stmt, error)
+	InsertUnsequencedLeafStmt(tx *sql.Tx) (*sql.Stmt, error)
+	InsertSequencedLeafStmt(tx *sql.Tx) (*sql.Stmt, error)
+	GetSequencedLeafCountStmt(tx *sql.Tx) (*sql.Stmt, error)
+}
+
+// MapStatementProvider provides SQL statement objects for map storage.
+type MapStatementProvider interface {
+	GetMapLeafStmt(tx *sql.Tx, num int) (*sql.Stmt, error)
+	GetLatestMapRootStmt(tx *sql.Tx) (*sql.Stmt, error)
+	InsertMapHeadStmt(tx *sql.Tx) (*sql.Stmt, error)
+	InsertMapLeafStmt(tx *sql.Tx) (*sql.Stmt, error)
+}
+
+// AdminStatementProvider provides SQL statement objects for administration.
+type AdminStatementProvider interface {
+	GetTreeIDsStmt(tx *sql.Tx) (*sql.Stmt, error)
+	GetAllTreesStmt(tx *sql.Tx) (*sql.Stmt, error)
+	GetTreeStmt(tx *sql.Tx) (*sql.Stmt, error)
+	InsertTreeStmt(tx *sql.Tx) (*sql.Stmt, error)
+	InsertTreeControlStmt(tx *sql.Tx) (*sql.Stmt, error)
+	UpdateTreeStmt(tx *sql.Tx) (*sql.Stmt, error)
+}
+
+// CustomBehaviourProvider abstracts database specific features, for example error code checking.
+type CustomBehaviourProvider interface {
+	CheckDatabaseAccessible(ctx context.Context) error
+	IsDuplicateErr(err error) bool
+	TreeRowExists(treeID int64) error
+}
+
+// LifecycleHooks allows implementations to add custom logic at various points in the
+// database and transaction flow.
+type LifecycleHooks interface {
+	OnOpenDB() error
+}
+
+// DBWrapper encapsulates a database and provides customized SQL statement objects for all types
+// of tree storage as well as abstracting some operations that differ between databases.
+// Statements returned by any of public functions belong to the caller transaction and must be
+// closed on completion of the work.
+type DBWrapper interface {
```

Comment:
For discussion:

A DBWrapper is a mix of "statement provider" (ie, a thing-that-has-queries) and a DAO (particularly for trees). Mixing different concepts usually makes things harder in the long run, as we're experiencing now.

What we're trying to do, as I understand, is allow support for distinct databases. The usual way to do that is via DAOs (or whatever other fancy name people use nowadays). The problem *we* have is that our storage layer mixes several different concepts and responsibilities into the storage and transaction objects.

A few of the issues are:
1. We represent collection of entities as "storages", instead of having per-entity DAOs
1. We confuse transactions with DAOs
1. We have too much tree-manipulation logic on the storage layer

The combination of those mean our "Storage" objects do too much work, do work in unexpected places (like TX.Begin() methods) and make it borderline impossible to have cross-"storage" transactions. Re-writing the storage layer to support different databases is cumbersome, complex, and error prone. We also don't provide interface-based test suites, so each implementor is on its own (and we can't assess quality uniformly between implementations).

A DAO-like approach, particularly with "dumb" DAOs, fixes all of those and allows for database-specific forks. We could even provide SQL-but-non-database-specific DAOs, into which one just plugs slightly different query strings, but with a dumb-DAO layer and a few utility methods that shouldn't be necessary (nor is much desirable, IMO). Transactions should be moved to method parameters, instead of the object used to manipulate entities.

I think we should strive to understand where we are, where we'd like to be (ideally), and what venues we can take to get as close as possible to that ideal. Since we felt the need to rework this layer, we could take the time to put together a proposal that moves us to a permanently better place.